### PR TITLE
[REVIEW] Fix MultiIndexes w/ datetime levels, ensure Column names are handled everywhere

### PR DESCRIPTION
**Pull request by trxcllnt**
**tldr;** This PR fixes https://github.com/rapidsai/cudf/issues/2523 and https://github.com/rapidsai/cudf/issues/2181. It started as a fix for #2523, but grew in scope to include #2181 as fixes uncovered subtler bugs/unimplemented features below.

An accounting of my journey since last Friday:

I thought fixing #2523 would be simple. Sometimes `MultiIndex._levels` contained Series, other times np.ndarrays or pd.Indexes. [Easy fix](https://github.com/rapidsai/cudf/commit/46b96dac0fb9919d5653887dc741eed5edebb156#diff-9074f390867cad6bd7558c315de67379R561-R563), add [a test](https://github.com/rapidsai/cudf/commit/047e3d041622f5e9a53e468f706a52ac55414134#diff-92763a1715f0ca6630e968f4090bb667R659), case closed.

But then I wonder if datetimes cause other issues for MultiIndexes. I didn't see tests explicitly testing datetimes, so I added [a datetime level](https://github.com/rapidsai/cudf/commit/e0453badc0eb26ad1aec5e3acf00c772d0a87049#diff-9173f3a49dccf990b251e7832ba6afdeR149) to the pytest fixture. And that's the moment all hell broke loose. `MultiIndex` names are missing, level dtypes are wrong, and I'm staring at a nightmare sea of pytest red.

The first broken test goes through the MultiIndex constructor, which [uses `join` to align the levels and codes](https://github.com/trxcllnt/cudf/blob/ae9b903ab4995ef756f5774d2749fad2835eed65/python/cudf/cudf/dataframe/multiindex.py#L103). But here, the join result dtype goes from `datetime64[ns]` to `datetime64[ms]`. Ah ha! I need to update `join` to [propagate the in-columns' `dtype_info`](https://github.com/rapidsai/cudf/commit/d160bb901a6936cc6dbd726dd501d86e3a078fae#diff-7321fb1b2faf444f9922620811887130R153) through to the out-columns. Easy fix, [add a test](https://github.com/rapidsai/cudf/commit/8ed2d981149c2f7c5e756e09cd7284039ebf9f24#diff-d84607ab4c6d441c3b0b3fbe876895dfR691), move on.

Prior experience with #2181 told me Series and Column names could get out of sync, so I intuited the same thing could be happening with Indexes. [Changing Indexes to forward calls on to their Column's `name`](https://github.com/rapidsai/cudf/commit/b5dcfa9146c22f6a8fd369d68a59f4f2fcc5a618) (and updating Series/Column to [propagate the name in `_replace_defaults`](https://github.com/rapidsai/cudf/commit/158f8ca1832324430de10803023aa72d862b5269#diff-e53ccb86a68d081a81e31cf27182d86cR377)) fixed both #2181 and the rest of the MultiIndex tests.

But now I've broken `DataFrame.drop_duplicates`, specifically where it calls into cython. The `in_index` name is correct going in, but the `new_index` name coming out is `None`:
https://github.com/rapidsai/cudf/blob/085bf2d4bc78882d8a0d38a3f39122d47485d45c/python/cudf/cudf/dataframe/dataframe.py#L1489-L1495

You may be wondering why I didn't just type `new_index.name = in_index.name` into my editor and end the nightmare right here. Clearly at some level I must enjoy pain, because instead of stopping I said to myself, "I've done all this work to ensure the Column knows its name, I'll be damned if cython is going to forget it now."

I see the `gdf_column` has a `char* col_name` field. I even find a few uses in cython. Unfortunately, it's only used in the avro, csv, json, and parquet readers/writers.

I'm going to have to touch nearly all the cython that creates or frees Columns, DataFrames, `gdf_columns`, and `cudf_tables`. Yet I persist. Erasing the name at this layer has probably been the source of python frustration and confusion, or it will be in the near future.

I add cython helper methods to [create cudf.Columns from gdf_columns](https://github.com/trxcllnt/cudf/blob/02fde3893c6bce9d9e81ffdff0dc2fc4c0681389/python/cudf/cudf/bindings/cudf_cpp.pyx#L275) and to [free `gdf_columns` and their related `col_name` pointers](https://github.com/trxcllnt/cudf/blob/02fde3893c6bce9d9e81ffdff0dc2fc4c0681389/python/cudf/cudf/bindings/cudf_cpp.pyx#L555). Eventually I compile, everything works, and I rest.

### CODE REVIEW

File: python/cudf/cudf/bindings/rolling.pyx [hunk 1]
```diff
@@ -68,16 +79,16 @@ def apply_rolling(inp, window, min_periods, center, op):
                 c_min_periods_col,
                 c_forward_window_col
             )
-        data, mask = gdf_column_to_column_mem(output_col)
-
-    result = Column.from_mem_views(data, mask)
+        # I'd expect this to work but it doesn't...
+        # result = gdf_column_to_column(c_output_col)
```
Comment (trxcllnt) [inline python/cudf/cudf/bindings/rolling.pyx]: Figuring this out was weird. I couldn't figure out what's different about this line vs. the two below it, since this function contains those two lines.

File: python/cudf/cudf/bindings/typecast.pyx [hunk 2]
```diff
@@ -33,15 +31,16 @@ def apply_cast(incol, dtype=np.float64):
...
             c_out_dtype,
             c_out_info
         )
 
-    free(c_incol)
-    data, mask = gdf_column_to_column_mem(&result)
-    return Column.from_mem_views(data, mask)
+    free_column(c_incol)
+
+    return gdf_column_to_column(&c_out_col)
```
Comment (trxcllnt) [inline python/cudf/cudf/bindings/typecast.pyx]: These changes and the others like it are updates to use the new helper method, and to apply a consistent style across similar call sites.

File: cpp/src/rolling/rolling.cu [hunk 3]
```diff
@@ -255,9 +256,17 @@ gdf_column* rolling_window(const gdf_column &input_col,
   // Use the column wrapper class from io/utilities to quickly create a column
   gdf_column_wrapper output_col(input_col.size,
                                 input_col.dtype,
-                                gdf_dtype_extra_info{TIME_UNIT_NONE},
+                                input_col.dtype_info,
                                 "");
 
+  if (input_col.col_name != nullptr) {
```
Comment (harrism) [inline cpp/src/rolling/rolling.cu]: This will be done by gdf_column_view_augmented when you pass a string name to gdf_column_wrapper constructor. So just change the creation of output_col above.

File: cpp/src/rolling/rolling.cu [hunk 4]
```diff
@@ -255,9 +256,17 @@ gdf_column* rolling_window(const gdf_column &input_col,
   // Use the column wrapper class from io/utilities to quickly create a column
   gdf_column_wrapper output_col(input_col.size,
                                 input_col.dtype,
-                                gdf_dtype_extra_info{TIME_UNIT_NONE},
+                                input_col.dtype_info,
                                 "");
```
Comment (harrism) [inline cpp/src/rolling/rolling.cu]:
```suggestion
                                input_col.col_name == nullptr ? "" : std::string(input_col.col_name);
```

File: python/cudf/cudf/bindings/rolling.pyx (hunk 1 above)
Comment (kkraus14) [reply python/cudf/cudf/bindings/rolling.pyx]: I think lets just open an issue and investigate post release, only thing I can think of is something weird with Cython scoping of Python objects between cdefs
